fix: Preview adjusts for device orientation change

dcalhoun opened the pull request:
Fixes #18483. After orientation changes, the device width may be too narrow for the existing preview, which causes it to overflow the view. This overflow is more prominent when Display Zoom is enabled, but is present regardless. By resetting the preview width after orientation change, we ensure the preview fits the view. 

<details><summary>Before</summary>

https://user-images.githubusercontent.com/438664/175136889-8e4b3c83-d266-4d49-b15b-9ec5c25353f1.mov

</details>

<details><summary>After</summary>

https://user-images.githubusercontent.com/438664/175137277-42547889-c3a7-4872-83ba-5f55c1d83832.mov

</details>

To test:

> **Note**: It is important to test on both an iPhone and iPad, as the reporting of view dimensions seems to vary between the device types. 

<details><summary>While previewing, rotate from portrait to landscape orientation</summary>

1. Open a published or drafted post.   
1. Position your device in _portrait_ orientation. 
1. Tap the three-dot menu in the top-right of the editor, select _Preview_.
1. Tap the device icon in the bottom-right and select a preview device size.
1. Rotate your device to _landscape_ orientation. 
1. Verify the preview does not overflow the width of the device; the entirety of the preview is visible on the screen.<sup>1</sup> 
1. Rotate your device to _portrait_ orientation. 
1. Verify the preview does not overflow the width of the device; the entirety of the preview is visible on the screen.<sup>1</sup> 
1. Repeat steps 4-8 with each of the preview device sizes available. 

<sup>1</sup> One way to verify the correct display is to again toggle the device away and back to your current preview device, e.g. "mobile" → "desktop" → "mobile" without rotating the device. This will re-render the preview and provide a clear picture as to what the preview should look like. 

</details>

<details><summary>While previewing, rotate from landscape to portrait orientation</summary>

1. Open a published or drafted post.   
1. Position your device in _landscape_ orientation. 
1. Tap the three-dot menu in the top-right of the editor, select _Preview_.
1. Tap the device icon in the bottom-right and select a preview device size.
1. Rotate your device to _portrait_ orientation. 
1. Verify the preview does not overflow the width of the device; the entirety of the preview is visible on the screen.<sup>1</sup> 
1. Rotate your device to _landscape_ orientation. 
1. Verify the preview does not overflow the width of the device; the entirety of the preview is visible on the screen.<sup>1</sup> 
1. Repeat steps 4-8 with each of the preview device sizes available. 

<sup>1</sup> One way to verify the correct display is to again toggle the device away and back to your current preview device, e.g. "mobile" → "desktop" → "mobile" without rotating the device. This will re-render the preview and provide a clear picture as to what the preview should look like. 

</details>

## Regression Notes
1. Potential unintended areas of impact
  The preview web view could experience bugs during rotation or memory leaks from the changes.
2. What I did to test those areas of impact (or what existing automated tests I relied on)
  Manually tested various devices, zoom settings, and multiple rotation events.
3. What automated tests I added (or what prevented me from doing so)
  None, I am interested in recommendations/guidance on if/what automated testing should be added.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dcalhoun committed:
fix: Preview adjusts for device orientation change and zoom

After orientation changes, the device width may be too narrow for the
existing preview, which causes it to overflow the view. By resetting the
preview width after orientation change, we ensure the preview fits the
view. Additionally, the Display Zoom setting is now taken into account
when computing the preview width.

dcalhoun committed:
docs: Add release note

twstokes commented on WordPress/Classes/ViewRelated/Post/Preview/PreviewWebKitViewController.swift:
```diff
@@ -104,6 +104,17 @@ class PreviewWebKitViewController: WebKitViewController {
+        NotificationCenter.default.addObserver(self, selector: #selector(orientatedChanged), name: UIDevice.orientationDidChangeNotification, object: nil)
+    }
+
+    deinit {
```
I _think_ this [isn't required](https://developer.apple.com/documentation/foundation/notificationcenter/1415360-addobserver) but it's one less thing for the system to clean up if we're explicit. 👍 

> If your app targets iOS 9.0 and later or macOS 10.11 and later, you do not need to unregister an observer that you created with this function. If you forget or are unable to remove an observer, the system cleans up the next time it would have posted to it.

twstokes reviewed (approved):
LGTM @dcalhoun, great work! 🚀 

Both tests succeeded for me on an iPhone 12:
- ✅ Change preview device while in landscape orientation
- ✅ Rotate device to landscape orientation with Display Zoom enabled

dcalhoun committed:
refactor: Improve function naming

dcalhoun committed:
fix: Correct width computation for tablets

The previous implementation resulted in inaccurate width computations.
The height/width dimensions can be reported in an unexpected, reversed
manner depending upon when they are accessed during an orientation
change event. Additionally, using the same computation for orientation
change and device selection change resulted in unexpected, different
results. Further researched revealed the `viewWillTransition` method,
which documentation states should be used "to make interface-based
adjustments."

`viewWillTransition` is now used to manage updates to the view size
(orientation change), and provides the incoming view dimensions to the
`setWidth` method.

- https://developer.apple.com/documentation/uikit/uiviewcontroller/1621373-interfaceorientation
- https://developer.apple.com/documentation/uikit/uicontentcontainer/1621466-viewwilltransition

dcalhoun committed:
refactor: Remove unnecessary calculation based upon Display Zoom

The scale calculation is unnecessary. The association of overflowing
content with Display Zoom was a "red herring" and it is unrelated. Now
that the preview correctly resizes during orientation changes, the
preview always fits the view size, regardless of whether Display Zoom is
enabled or not.

dcalhoun committed:
docs: Remove Display Zoom reference from release note

The bug is unrelated to Display Zoom.

dcalhoun commented:
@twstokes after further testing, I identified a scenario where overflowing content still occurred. Namely, when using an iPad Pro 11 inch in landscape orientation, selecting "Desktop" as the preview device, and rotating to portrait orientation. 

Further debugging and research led me to believe the original issue is solely related to orientation changes and unrelated to Display Zoom. I have updated the implementation to use [`viewWillTransition`](https://developer.apple.com/documentation/uikit/uicontentcontainer/1621466-viewwilltransition), which I believe is the most appropriate method. I have also removed previous code attempting to detect the Display Zoom scale. 

I'd appreciate your review again. I updated the PR description and testing instructions to be more robust and provided before/after videos as well.  Thank you for the help and your patience with a bit of back-and-forth. 🙇🏻

twstokes commented on WordPress/Classes/Utility/WebKitViewController.swift:
```diff
@@ -359,9 +359,18 @@ class WebKitViewController: UIViewController, WebKitAuthenticatable {
+    func setWidth(_ width: CGFloat?, viewSize: CGSize? = nil) {
         if let width = width {
-            widthConstraint?.constant = min(width, view.superview?.frame.width ?? width)
+            var horizontalViewBound: CGFloat
```
Nit: This could be a constant.

```suggestion
            let horizontalViewBound: CGFloat
```

twstokes commented on WordPress/Classes/Utility/WebKitViewController.swift:
```diff
@@ -359,9 +359,18 @@ class WebKitViewController: UIViewController, WebKitAuthenticatable {
 
     /// Sets the width of the web preview
     /// - Parameter width: The width value to set the webView to
```
Could we add a parameter comment for `viewSize` here? 🙇

twstokes commented on WordPress/Classes/Utility/WebKitViewController.swift:
```diff
@@ -359,9 +359,18 @@ class WebKitViewController: UIViewController, WebKitAuthenticatable {
+            var horizontalViewBound: CGFloat
+            if let viewWidth = viewSize?.width {
+                horizontalViewBound = viewWidth
+            } else if let superViewWidth = view.superview?.frame.width {
```
I realized that `view.superview?.frame.width` won't have the "going to" value from `viewWillTransition`, otherwise we could simplify things some. This makes sense. 👍

dcalhoun replied:
I interpret your comment to mean no changes are needed here. Please correct me if I am mistaken, however.

twstokes replied:
Yep! All good. 👍

twstokes commented on WordPress/Classes/ViewRelated/Post/Preview/PreviewWebKitViewController.swift:
```diff
@@ -106,6 +106,11 @@ class PreviewWebKitViewController: WebKitViewController {
         setupDeviceLabel()
     }
 
+    override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
```
Good call going to this class method over the notification. It's a good fit for this case. 👍

twstokes commented on WordPress/Classes/Utility/WebKitViewController.swift:
```diff
@@ -359,9 +359,18 @@ class WebKitViewController: UIViewController, WebKitAuthenticatable {
     /// Sets the width of the web preview
     /// - Parameter width: The width value to set the webView to
-    func setWidth(_ width: CGFloat?) {
+    func setWidth(_ width: CGFloat?, viewSize: CGSize? = nil) {
```
Nit: If we only need the width from `viewSize`, maybe we should just pass that in?

twstokes reviewed (commented):
👋 @dcalhoun! Thanks for all your work on this. I've made some comments on the PR and will do device testing now. 🙇

dcalhoun committed:
refactor: Prefer constant over variable

Co-authored-by: Tanner Stokes <tanner.stokes@automattic.com>

dcalhoun committed:
refactor: Simplify, document Preview orientation change implementation

Reduce function arguments to minimum required values. Update existing
function documentation to include new arguments.

dcalhoun committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-iOS into fix/preview-adjusts-for-device-orientation-change-and-zoom

twstokes commented:
I tested on an iPhone 12 with and without Display Zoom enabled, as well as an iPad Pro in Simulator and did not spot any issues. 🚀 

✅ While previewing, rotate from portrait to landscape orientation
✅ While previewing, rotate from landscape to portrait orientation
